uportal-war/src/main/java/org/jasig/portal/layout/dlm/remoting/UpdatePreferencesServlet.java

```diff
@@ -863,7 +882,133 @@ public ModelAndView addTab(HttpServletRequest request, HttpServletResponse respo
     }
+    
+    /**
+     * Add a new folder to the layout.
+     * 
+     * @param request
+     * @param response
+     * @param targetId - id of the node to add the new folder to. By default, the folder will be inserted after other
+     *                   existing items in the node unless a siblingId is provided.
+     * @param siblingId - if set, insert new folder prior to the node with this id, otherwise simple insert at the end of the list.
+     * @param attributes - if included, parse the JSON name-value pairs in the body as the attributes of the folder.
+     * e.g. :
+     * {   
+     *      "structureAttributes" : {"display" : "row", "other" : "another" },
+     *      "attributes" : {"hidden": "true", "type" : "header-top" }
+     * }
+     * 
+     * @return
+     */
+    @RequestMapping(method = RequestMethod.POST, params = "action=addFolder")
+    public ModelAndView addFolder(HttpServletRequest request, 
+                                  HttpServletResponse response, 
+                                  @RequestParam("targetId") String targetId, 
+                                  @RequestParam(value="siblingId", required=false) String siblingId,
+                                  @RequestParam(value="display", required=false) String display,
+                                  @RequestBody(required=false) Map<String, Map<String, String>> attributes) {
+        IUserLayoutManager ulm = userInstanceManager.getUserInstance(request).getPreferencesManager().getUserLayoutManager();
+
+        if (!ulm.getNode(targetId).isAddChildAllowed()) {
+            response.setStatus(403);
+            return null;
+        }
 
+        UserLayoutFolderDescription newFolder = new UserLayoutFolderDescription();
+        newFolder.setHidden(false);
+        newFolder.setImmutable(false);
+        newFolder.setAddChildAllowed(true);
+        newFolder.setFolderType(IUserLayoutFolderDescription.REGULAR_TYPE);
+        
+        // Update the attributes based on the supplied JSON (optional request body name-value pairs)
+        if (attributes != null && !attributes.isEmpty()) {
+            setObjectAttributes(newFolder, request, attributes);
+        }
+
+        ulm.addNode(newFolder, targetId, siblingId);
+        final Locale locale = RequestContextUtils.getLocale(request);
+        
+        try {
+            ulm.saveUserLayout();
+        } catch (Exception e) {
+            log.warn("Error saving layout", e);
+            return new ModelAndView("jsonView", Collections.singletonMap("error", getMessage("error.persisting.layout.change.folder", "Unable to add a new folder", locale)));
+        }        
+
+        Map<String, Object> model = new HashMap<>();
+        model.put("response", getMessage("success.add.folder", "Added a new folder", locale));
+        model.put("folderId", newFolder.getId());
+        model.put("immutable", newFolder.isImmutable());
```

Comment:
You don't need to change the code, but FYI the immutable and unremovable attributes are somewhat obsolete.  I did note that the Respondr theme still uses them somewhat, but they have been replaced by the dlm:addChildAllowed, dlm:deleteAllowed, dlm:editAllowed, and dlm:moveAllowed . See https://wiki.jasig.org/display/UPM42/Setup+a+fragment+layout+by+modifying+fragment-layout+XML+files.